### PR TITLE
Datagen: Cache supported_locales impls

### PR DESCRIPTION
**Pull request by sffc**
`cargo make bakeddata datetime` performance on my machine:

On main: 173.91 seconds.

On this branch: 62.24 seconds.

Still slow but it seems to be substantially faster.

### CODE REVIEW

**Comment by sffc** [comment]
Based on the flamegraph, it seems after this lands the next bottlenecks might be:

1. `impl Display for TokenStream` (this seems to be by far the biggest contributor now)
2. rustfmt (seems to be almost half of the time spent??)
3. `impl Bake for DateSkeletonPatternsV1` (small but big enough to show up)

File: provider/datagen/src/provider.rs [hunk 1]
```diff
@@ -306,6 +309,28 @@ pub struct SourceData {
     pub(crate) icuexport_dictionary_fallback: Option<Arc<SerdeCache>>,
     #[cfg(feature = "legacy_api")]
     pub(crate) collations: Vec<String>,
+    pub(crate) supported_locales_cache_vec:
```
Comment (robertbastian) [inline provider/datagen/src/provider.rs:312]: Use an `Arc`?

File: provider/datagen/src/provider.rs [hunk 2]
```diff
@@ -485,3 +510,34 @@ impl SourceData {
...
+        self.source
+            .supported_locales_cache_vec
+            .0
+            .insert_with(M::KEY, || Box::from(self.supported_locales_impl()))
+            .as_ref()
+            .cloned()
+            .map_err(|e| *e)
+    }
+
+    fn supports_locale(&self, locale: &DataLocale) -> Result<bool, DataError> {
```
Comment (robertbastian) [inline provider/datagen/src/provider.rs:533]: `supports_locale` is called n times per key, whereas `supported_locales` is only called once. It probably makes more sense to cache a `HashSet` and convert it to a `Vec` for the one call, than to linearly scan a `Vec` n times.

**Review by robertbastian — COMMENTED** [review]
observation: `DatagenDriver` already calls `supported_locales` exactly once per key, so this is a `DatagenProvider`-internal issue.

**Comment by Manishearth** [comment]
> rustfmt (seems to be almost half of the time spent??)


n.b. we don't run this in google3 anyway

File: provider/datagen/src/provider.rs [hunk 3]
```diff
@@ -306,6 +309,28 @@ pub struct SourceData {
     pub(crate) icuexport_dictionary_fallback: Option<Arc<SerdeCache>>,
     #[cfg(feature = "legacy_api")]
     pub(crate) collations: Vec<String>,
+    pub(crate) supported_locales_cache_vec:
+        FrozenMapThrowawayClone<DataKey, Box<Result<Vec<DataLocale>, DataError>>>,
+}
+
+pub(crate) struct FrozenMapThrowawayClone<K, V>(pub(crate) FrozenMap<K, V>);
```
Comment (Manishearth) [inline provider/datagen/src/provider.rs:316]: I believe `elsa` master has these impls now

Comment (Manishearth) [reply provider/datagen/src/provider.rs:316]: It does for `sync` at least.

Comment (Manishearth) [reply provider/datagen/src/provider.rs:316]: `elsa 1.10` should work.

**Review by Manishearth — COMMENTED** [review]
r=me with the change from vec to hashset

File: provider/datagen/src/provider.rs (hunk 2 above)
Comment (sffc) [reply provider/datagen/src/provider.rs:533]: Performance on `cargo make bakeddata datetime` (run time only, not build time):

| Commit | Trial 1 | Trial 2 | Trial 3 |
|---|---|---|---|
| 11a1ed1 (Vec) | 53.18 | 61.74 | 53.82 |
| f2de286 (HashSet) | 58.64 | 51.89 | 51.03 |

So it looks like HashSet wins.

File: provider/datagen/src/provider.rs (hunk 1 above)
Comment (sffc) [reply provider/datagen/src/provider.rs]: Switched to Arc

File: provider/datagen/src/provider.rs (hunk 3 above)
Comment (sffc) [reply provider/datagen/src/provider.rs]: Upgraded to elsa 1.10

File: provider/datagen/src/provider.rs [hunk 4]
```diff
@@ -520,24 +519,11 @@ where
...
-        #[allow(deprecated)] // SourceData
-        self.source
-            .supported_locales_cache_vec
-            .0
-            .insert_with(M::KEY, || Box::from(self.supported_locales_impl()))
-            .as_ref()
-            .cloned()
-            .map_err(|e| *e)
+        self.supported_locales_vec()
+            .map(|v| v.iter().cloned().collect())
```
[the remark this answers is not in the document]
Comment (sffc) [reply provider/datagen/src/provider.rs:524]: Yes thanks fixed

**Review by robertbastian — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Comment by sffc** [comment]
Changelog in https://github.com/unicode-org/icu4x/pull/4484

**Pull request merged**